content/english/java-tictactoe/activity-5.md

```diff
@@ -1,7 +1,111 @@
+
+In this exercise, we want to increase the difficulty of the game by making optimal decisions for the computer.
+
+We will do so by adding artificial intelligence into our program with the use the <b>Minimax</b> algorithm (a well-defined procedure that allows computers to solve problems).
+
+### Minimax Algorithm
+Minimax is an algorithm that is used in two-player games to make optimal decisions for a player. 
+
+- The algorithm makes optimal decision for the maximizer by examining all the possible future game states based on the current game board. 
+- The two players are labeled as <b>maximizer</b> and <b>minimizer</b> respectedly. In this case, the computer is the maximizer, and the player is the minimizer. 
+- If there is a winner or tie, that game state will be given a score.
+- In our game specifically, we want the computer to win in fewest steps possible. Hence, we design the calculation of score for the game board as the following: 
+
+   (computer wins (`1 * (number of spots left + 1)`), player wins (`-1 * (number of spots left + 1)`), tie (`0`)).
+- While the maximizer tries to get the highest score possible, the minimizer tries to minimize the score of its opponent.  
+
+Let's look at an example below:
+
+<img src="../images/minimax.png" height="500"/> 
+
+1. In the first row, we considered the 3 possible moves for `"O"`. 
+
+2. We note that if the computer puts its move at position 8, the game will halt as the computer wins. That state will hence be given a score of `1 * (number of spots left + 1)` = `8`.
+
+3. We would consider all the possible moves of `"X"` and `"O"` until all possibilities end with a winner or a tie.
+
+4. At game states that don't have a winner or a tie, we pick the smallest score during minimizing rounds, and the largest score during maximizing rounds.
+
+If you follow along through the maximizing/minimizing rounds on the picture above, you should note that the 3 possible moves for "O" at the starting game board have the scores of `0`(position 7), `3`(position 8), `0`(position 6). 
+
+Hence, the optimal move for the computer is to place `"O"` at position 8.
+
+### Write the Method `getComputerMoveAI()`
+In activity-3, you wrote the method `int getComputerMove(String[] curBoard)` to randomly generate a spot for the computer. Let's write another method called `getComputerMoveAI(String[] curBoard)` that returns the optimal move for the computer. 
+
+We will write this method assuming we have the `int minimax(String[] curBoard, boolean isMaximizing)` method implemented, which you will write later.
+
+1. For each available spot on the game board, place `"O"` at that spot and get the score for that board by calling `minimax()`. 
+
+{{% notice hint %}}
+- Note that we call the method `minimax()` inside of the implementation of itself. We call a function that calls itself during its execution a <b>recursive function</b>.
```

Comment:
Have we talked about recursion before? If not, we might have to add an advanced concept under function in basics